<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="Piper : piRNA pipeline developed in the Zamore Lab and ZLab in UMass Med School" />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Piper</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/bowhan/Piper">View on GitHub</a>

          <h1 id="project_title">Piper</h1>
          <h2 id="project_tagline">piRNA pipeline developed in the Zamore Lab and ZLab in UMass Med School</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/bowhan/Piper/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/bowhan/Piper/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <p><img src="https://dl.dropboxusercontent.com/u/5238651/pipipe%20logo.jpg" alt="LOGO" title="RTI UMASSMED HHMI"></p>

<h1>
<a name="piper" class="anchor" href="#piper"><span class="octicon octicon-link"></span></a>Piper</h1>

<p>Integrated pipeline collections developed in the <a href="http://www.umassmed.edu/zamore">Zamore Lab</a> and <a href="http://zlab.umassmed.edu/zlab">ZLab</a> to analyze piRNA/transposon for different Next Generation Sequencing (<em>small RNA-Seq, RNA-Seq, Genomic-Seq, ChIP-Seq, CAGE-Seq and Degradome-Seq</em>). <strong>piper</strong> provides a generic interface for different organisms/genomes with a particular optimization for fruit-fly and mouse, which are the main focus in our labs as well as the piRNA field.</p>

<p>For <em>small RNA-Seq</em>, <em>RNA-Seq</em> and <em>ChIP-Seq</em> pipelines, <strong>piper</strong> provides two modes: <code>a single library mode</code> and <code>dual library mode</code>, to analyze single library and pair-wise comparison between two samples respectively.     </p>

<h2>
<a name="install" class="anchor" href="#install"><span class="octicon octicon-link"></span></a>INSTALL</h2>

<p><strong>piper</strong> is written in Bash, C++, Perl, Python and R. It currently only works under Linux environment.</p>

<p><strong>piper</strong> uses the following public tools:</p>

<ol>
<li><p>For alignment, <strong>piper</strong> uses <a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a>, <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a>, <a href="http://bio-bwa.sourceforge.net/">BWA</a>,  <a href="https://code.google.com/p/rna-star/">STAR</a> and <a href="http://mrfast.sourceforge.net/">mrFast</a> for different purposes.       </p></li>
<li><p>For transcripts/transposons quantification, <strong>piper</strong> uses <a href="http://cufflinks.cbcb.umd.edu/">Cufflinks</a>, <a href="http://www-huber.embl.de/users/anders/HTSeq/">HTSeq</a> and <a href="http://bio.math.berkeley.edu/eXpress">eXpress</a> under different circumstances.     </p></li>
<li><p>For transposon mobilization discovery, <strong>piper</strong> uses <a href="http://zlab.umassmed.edu/%7Ezhuangj/TEMP/">TEMP</a>, <a href="http://gmt.genome.wustl.edu/breakdancer/current/">BreakDancer</a>, <a href="https://github.com/tk2/RetroSeq">RetroSeq</a> and <a href="http://compbio.cs.sfu.ca/software-variation-hunter">VariationHunter</a>.    </p></li>
<li><p>For peaks calling for ChIP-Seq, <strong>piper</strong> uses <a href="https://github.com/taoliu/MACS">MACS2</a>.    </p></li>
<li><p>Additionally, <strong>piper</strong> uses many tools from the <a href="http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64.v287/">Kent Tools</a>, like <code>faSize</code>, <code>bedGraphToBigWig</code>.   </p></li>
<li><p>To wrap bash scripts for multi-threading, <strong>piper</strong> utilizes <code>ParaFly</code> from <a href="http://trinityrnaseq.sourceforge.net/">Trinity</a>. <strong>piper</strong> also learns the <code>touch</code> trick for job resuming from <a href="http://trinityrnaseq.sourceforge.net/">Trinity</a>.        </p></li>
<li><p>To determine the version of FastQ, <strong>piper</strong> uses a modified version of <code>SolexaQA.pl</code> from <a href="http://solexaqa.sourceforge.net/">SolexaQA</a>.     </p></li>
<li><p><strong>piper</strong> uses <a href="https://github.com/bowhan/bedtools.git">BEDtools</a> with slight modification on <code>intersectBed.cpp</code> to accommodate a special BED format used in the pipeline. It ships with the modified source code, as well as statically compiled binary renamed as <code>bedtools_piper</code> to avoid confusion with the real one.       </p></li>
</ol><hr><h3>
<a name="cc" class="anchor" href="#cc"><span class="octicon octicon-link"></span></a>C/C++</h3>

<p><strong>piper</strong> ships with statically compiled linux x86_64 binaries for its own C++ codes and all the other tools written in C/C++. Ideally, the users don't need to do any compiling. But if the static versions do not work in your system, please install them and move the binaries to the <code>bin</code>. For BEDtools, please install the one in the <code>third_party</code> directory and rename it as <code>bedtools_piper</code> in the <code>bin</code> directory of <code>piper</code> For your convenience, the source codes of all C/C++ tools have been included as tar ball. Some of piper C++ codes utilizes <em>C++11</em> features and <em>Boost</em> libraries. It is recommended to install relatively new <a href="http://gcc.gnu.org/">GCC</a> and <a href="http://www.boost.org/users/download/">Boost</a> if compiling needs to be done.      </p>

<hr><h3>
<a name="pythoncython" class="anchor" href="#pythoncython"><span class="octicon octicon-link"></span></a>Python/Cython</h3>

<p><strong>For MACS[8] and HTSeq-count[13], the users will need to install them and make them available in the <code>$PATH</code>.</strong><br><em>We cannot find a good way to ship the ready-to-use Cython code. Without <code>htseq-count</code>, <code>piper rna/deg/cage</code> won't be able to make transcripts/transposon counting using genomic coordinates. But it will still perform other functions of the pipeline as well as quantification using Cufflinks and eXpress. Without <code>macs2</code>, <code>piper chip/chip2</code> won't work at all</em>.</p>

<hr><h3>
<a name="r" class="anchor" href="#r"><span class="octicon octicon-link"></span></a>R</h3>

<p>For R packages that are unavailable in the user's system, the installation is performed in the <code>piper install</code> process. They will be installed in the same directory as the pipeline in case the user doesn't have write permission in the R installation directory.</p>

<hr><h3>
<a name="genome-annotation" class="anchor" href="#genome-annotation"><span class="octicon octicon-link"></span></a>Genome Annotation</h3>

<p>Due to the limitation on the size of the files on github, the genome sequence, annotation files are to be downloaded from somewhere else and reformatted to accommodate the pipeline. <strong>piper</strong> uses <a href="http://support.illumina.com/sequencing/sequencing_software/igenome.ilmn">iGenome</a> and provides <code>piper install</code> to download iGenome genomes and organize the files to be used by the pipeline (see below).     </p>

<hr><h2>
<a name="usage" class="anchor" href="#usage"><span class="octicon octicon-link"></span></a>USAGE</h2>

<p>The pipeline finds almost everything under its own directory so please do not move the <code>zpipe</code> script. Use <code>ln  -s  $PATH_TO_piper/piper  $HOME/bin/piper</code> to create symbol link in your <code>$HOME/bin</code>; Or add <code>$PATH_TO_piper</code> to your <code>$PATH</code>. 
<strong>But please do not add the <code>$PATH_TO_piper/bin</code> to your <code>$PATH</code></strong></p>

<p>Call different pipelines using:     </p>

<div class="highlight highlight-Bash"><pre><span class="c"># genome installation pipeline</span>
 <span class="c"># 1. to install genome and R packages in one step</span>
<span class="nv">$PATH_TO_piper</span>/piper    install -g dm3<span class="p">|</span>mm9<span class="p">|</span>hg19... 
 <span class="c"># 2. to only download the genome and install R packages (if the node is not appropriate to be used for building indexes); then run (1).</span>
<span class="nv">$PATH_TO_piper</span>/piper    install -g dm3<span class="p">|</span>mm9<span class="p">|</span>hg19 -D
 <span class="c"># 3. to only download the iGenome that piper doesn't know (yet)</span>
<span class="nv">$PATH_TO_piper</span>/piper    install -g hg18 -l ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Homo_sapiens/UCSC/hg18/Homo_sapiens_UCSC_hg18.tar.gz

<span class="c"># to run small RNA pipeline in single library mode; input fastq can be gzipped</span>
<span class="nv">$PATH_TO_piper</span>/piper    small -i input.trimmed.fq<span class="o">[</span>.gz<span class="o">]</span> -g dm3 -c 24

<span class="c"># to run small RNA pipeline in dual library mode (need single library mode been ran for each sample first)</span>
<span class="nv">$PATH_TO_piper</span>/piper    small2 -a directory_A -b directory_B -g dm3 -c 24
<span class="c"># to run small RNA pipeline in dual library mode, normalized to miRNA </span>
<span class="nv">$PATH_TO_piper</span>/piper    small2 -a directory_A -b directory_B -g mm9 -c 24 -N miRNA
<span class="c"># to run small RNA pipeline in dual library mode, normalized to siRNA (structural loci and cis-NATs), for oxidation sample of -fruitfly only-</span>
<span class="nv">$PATH_TO_piper</span>/piper    small2 -a directory_A -b directory_B -g dm3 -c 24 -N siRNA

<span class="c"># to run RNASeq pipeline in single library mode, dUTP based method</span>
<span class="nv">$PATH_TO_piper</span>/piper    rnaseq -l left.fq -r right.fq -g mm9 -c 8 -o output_dir

<span class="c"># to run RNASeq pipeline in single library mode, ligation based method</span>
<span class="nv">$PATH_TO_piper</span>/piper    rnaseq -l left.fq -r right.fq -g mm9 -c 8 -o output_dir -L

<span class="c"># to run RNASeq pipeline in dual library mode (need single library mode been ran for each sample first)</span>
<span class="nv">$PATH_TO_piper</span>/piper    rnaseq2 -a directory_A -b directory_B -g mm9 -c 8 -o output_dir -A w1 -B piwi

<span class="c"># to run Degradome/RACE/CAGE-Seq library </span>
<span class="nv">$PATH_TO_piper</span>/piper    deg -l left.fq -r right.fq -g dm3 -c 12 -o output_dir

<span class="c"># to run ChIP Seq library in single library mode, for narrow peak, like transcriptional factor</span>
<span class="nv">$PATH_TO_piper</span>/piper    chip -l left.IP.fq -r right.IP.fq -L left.INPUT.fq -R right.INPUT.fq -g mm9 -c 8 -o output_dir

<span class="c"># to run ChIP Seq library in single library mode, for broad peak, like H3K9me3</span>
<span class="nv">$PATH_TO_piper</span>/piper    chip -l left.IP.fq -r right.IP.fq -L left.INPUT.fq -R right.INPUT.fq -g mm9 -c 8 -o output_dir -B

<span class="c"># to run ChIP Seq library in single library mode, only use unique mappers (otherwise Bowtie2 randomly choose one best alignment for each read)</span>
<span class="nv">$PATH_TO_piper</span>/piper    chip -l left.IP.fq -r right.IP.fq -L left.INPUT.fq -R right.INPUT.fq -g mm9 -c 8 -o output_dir -Q 10

<span class="c"># to run ChIP Seq library in dual library mode (need single library mode been ran for each sample first)</span>
<span class="nv">$PATH_TO_piper</span>/piper    chip2 -a directory_A -b directory_B -g mm9 -c 8 -o output_dir

<span class="c"># to run ChIP Seq library in dual library mode, extend up/down stream 5000 bp for TSS/TES/meta analysis (by bwtool)</span>
<span class="nv">$PATH_TO_piper</span>/piper    chip2 -a directory_A -b directory_B -g mm9 -c 8 -o output_dir -x 5000

<span class="c"># to run Genome Seq library</span>
<span class="nv">$PATH_TO_piper</span>/piper    dna -l left.fq -r right.fq -g dm3 -c 24 -D 100
</pre></div>

<p>Find more detailed information on <a href="https://github.com/bowhan/piper/wiki">Wiki</a></p>

<h3>
<a name="install-to-install-genome-assembly" class="anchor" href="#install-to-install-genome-assembly"><span class="octicon octicon-link"></span></a><em>install</em>: to install genome assembly</h3>

<p>Due to the limitation on the size of file by github, piper doesn't ship with the 
genome sequences and annotation. Alternatively, we provide scrips 
to download genome assemly files from iGenome project of illumina. Please make 
sure internet is available during this process.  <strong>piper</strong> provides an option to separate downloading from other processes, in case the machine/node with internet access is not appropriate for building index and other works.<br>
Except for the genome, this pipeline will also install unavailable R packages 
under the pipeline directory. The downloading and installation can be separated using -D option, in case the head node is not supposed to be used for heavy computational work, like building indexes.      </p>

<p>A more detailed explanation can be found <a href="https://github.com/bowhan/piper/wiki/installation">here</a>.  </p>

<h3>
<a name="small-small-rna-pipeline" class="anchor" href="#small-small-rna-pipeline"><span class="octicon octicon-link"></span></a><em>small</em>: small RNA pipeline</h3>

<p>small RNA library typically clones 18â€“40nt small RNAs, including miRNA, siRNA and 
piRNA. This pipeline maps those reads to rRNA, microRNA hairpin, genome, repbase 
annotated  transposons, piRNA clusters with bowtie and uses bedtools to assign 
them to different annotations. For each feature, length distribution, nucleotide percentage, 
ping-pong score, et al.,  are calculated and graphed. Some microRNA analysis is also included. 
In the dual library mode, pair-wise comparison of miRNA and piRNAs will be done. We invented this balloon-plot to efficiently compare the heterogeneity of miRNA between two samples. piRNA for different transposon family is also compared. </p>

<p>A more detailed explanation can be found <a href="https://github.com/bowhan/piper/wiki/smallRNA">here</a>.</p>

<h3>
<a name="rnaseq-rnaseq-pipeline" class="anchor" href="#rnaseq-rnaseq-pipeline"><span class="octicon octicon-link"></span></a><em>rnaseq</em>: RNASeq pipeline</h3>

<p>RNASeq pipeline can be used for both dUTR or ligation based RNASeq. 
It uses bowtie2 to aligns paired-end reads to rRNA and STAR[4] to align the unmapped reads 
to genome; Then it uses Cufflinks for quantification of transcripts from the genomic coordinates. It
also use HTSeq-count to quantify genomic features using coordinates. It also directly aligns reads to 
transcriptome, repbase annotated transposon, piRNA clusters using Bowtie2. Quantification 
was done using eXpress. Library is normalized by gene transcriptome compatible reads, given by Cufflinks. Basic statistics and graphs will be given. </p>

<p>A more detailed explanation can be found <a href="https://github.com/bowhan/piper/wiki/RNASeq">here</a>.  </p>

<h3>
<a name="cagedeg-cage--degradome-pipeline" class="anchor" href="#cagedeg-cage--degradome-pipeline"><span class="octicon octicon-link"></span></a><em>cage/deg</em>: CAGE &amp; Degradome pipeline</h3>

<p>Both types of libraries are designed to gather the information of the 5' end of RNAs 
CAGE clones RNAs with Cap and Degradome clones RNAs with 5' monophosphate. 
The pipeline will align reads to rRNA with bowtie2, genome using STAR. 
Different from RNASeq, this pipeline emphasizes the accuracy of the 5' ends. Nucleotide
composition surrounding the 5' end of the reads are given, like in small RNA library.</p>

<p>A more detailed explanation can be found <a href="https://github.com/bowhan/piper/wiki/DegradomeSeq">here</a>.</p>

<h3>
<a name="chip-chip-seq-pipeline" class="anchor" href="#chip-chip-seq-pipeline"><span class="octicon octicon-link"></span></a><em>chip</em>: ChIP-Seq pipeline</h3>

<p>ChIP Seq pipeline aligns both input and ChIP data to genome with Bowtie2. Peak calling was done
using MASC2. Signal is normalized in three different methods (ppois, FE and logLR). TSS/TES/meta plots are drawn using bwtool. In the dual library mode, peak calling is redone for each sample without inter-library normalization, by differential peak calling algorithm of MACS2 directly. TSS/TES/meta plots are drawn for those loci using the normalized signal. </p>

<p>A more detailed explanation can be found <a href="https://github.com/bowhan/piper/wiki/ChIPSeq">here</a>. </p>

<h3>
<a name="dna-genomic-seq-pipeline" class="anchor" href="#dna-genomic-seq-pipeline"><span class="octicon octicon-link"></span></a><em>dna</em>: Genomic Seq pipeline</h3>

<p>Genomic Seq pipelines aligns the paired-end reads to genome with Bowtie2, BWA-MEM and mrFast. Variations
were called using different algorithms. </p>

<p>A more detailed explanation can be found <a href="https://github.com/bowhan/piper/wiki/GenomeSeq">here</a>.</p>

<h2>
<a name="citing" class="anchor" href="#citing"><span class="octicon octicon-link"></span></a>Citing</h2>

<ul>
<li>in preparation</li>
</ul><h2>
<a name="contact" class="anchor" href="#contact"><span class="octicon octicon-link"></span></a>Contact</h2>

<pre><code>Wei Wang (wei.wang2 `at` umassmed.edu)
Bo W Han (bo.han `at` umassmed.edu, bowhan `at` me.com)
</code></pre>

<h2>
<a name="references" class="anchor" href="#references"><span class="octicon octicon-link"></span></a>References</h2>

<pre><code>[1] Li H and Durbin R. 2009. Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics 25: 1754-1760.  
[2] Langmead B, Trapnell C, Pop M and Salzberg SL. 2009. Ultrafast and memory-efficient alignment of short DNA sequences to the human genome. Genome Biol 10: R25.  
[3] Langmead B and Salzberg SL. 2012. Fast gapped-read alignment with Bowtie 2. Nat Methods 9: 357-359.    
[4] Dobin A, Davis CA, Schlesinger F, Drenkow J, Zaleski C, Jha S, Batut P, Chaisson M and Gingeras TR. 2013. STAR: ultrafast universal RNA-seq aligner. Bioinformatics 29: 15-21.    
[5] Trapnell C, Williams BA, Pertea G, Mortazavi A, Kwan G, van Baren MJ, Salzberg SL, Wold BJ and Pachter L. 2010.     
[6] Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation. Nat Biotechnol 28: 511-515. 
[7] Kent WJ, Zweig AS, Barber G, Hinrichs AS and Karolchik D. 2010. BigWig and BigBed: enabling browsing of large distributed datasets. Bioinformatics 26: 2204-2207.     
[8] Zhang Y et al. 2008. Model-based analysis of ChIP-Seq (MACS). Genome Biol 9: R137.  
[9] Roberts A and Pachter L. 2013. Streaming fragment assignment for real-time analysis of sequencing experiments. Nat Methods 10: 71-73.  
[10] Grabherr MG et al. 2011. Full-length transcriptome assembly from RNA-Seq data without a reference genome. Nat Biotechnol 29: 644-652.     
[11] Karolchik D, Hinrichs AS, Furey TS, Roskin KM, Sugnet CW, Haussler D and Kent WJ. 2004. The UCSC Table Browser data retrieval tool. Nucleic Acids Res 32: D493-D496.       
[12] Cox MP, Peterson DA and Biggs PJ. 2010. SolexaQA: At-a-glance quality assessment of Illumina second-generation sequencing data. BMC Bioinformatics 11: 485.    
[13] HTSeq: Analysing high-throughput sequencing data with Python. [http://www-huber.embl.de/users/anders/HTSeq/]  
[14] Keane TM, Wong K and Adams DJ. 2013. RetroSeq: transposable element discovery from next-generation sequencing data. Bioinformatics 29: 389-390.    
[15] Hormozdiari F, Hajirasouliha I, Dao P, Hach F, Yorukoglu D, Alkan C, Eichler EE and Sahinalp SC. 2010. Next-generation VariationHunter: combinatorial algorithms for transposon insertion discovery. Bioinformatics 26: i350-i357.    
</code></pre>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Piper maintained by <a href="https://github.com/bowhan">bowhan</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
